browser/ui/webui/federated_learning_internals_ui.cc

```diff
@@ -0,0 +1,128 @@
+/* Copyright (c) 2021 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/browser/ui/webui/federated_learning_internals_ui.h"
+
+#include <memory>
+#include <utility>
+#include <vector>
+
+#include "base/bind.h"
+#include "base/callback_helpers.h"
+#include "base/values.h"
+#include "brave/browser/brave_federated/brave_federated_service_factory.h"
+#include "brave/browser/ui/webui/brave_webui_source.h"
+#include "brave/common/webui_url_constants.h"
+#include "brave/components/brave_federated/brave_federated_service.h"
+#include "brave/components/brave_federated/data_store_service.h"
+#include "brave/components/brave_federated/data_stores/ad_notification_timing_data_store.h"
+#include "chrome/browser/profiles/profile.h"
+#include "chrome/grit/browser_resources.h"
+#include "chrome/grit/generated_resources.h"
+#include "components/grit/brave_components_resources.h"
+#include "content/public/browser/browser_context.h"
+#include "content/public/browser/web_contents.h"
+#include "content/public/browser/web_ui.h"
+#include "content/public/browser/web_ui_data_source.h"
+
+FederatedLearningInternalsHandler::FederatedLearningInternalsHandler(
+    Profile* profile) {
+  data_store_service_ =
+      brave_federated::BraveFederatedServiceFactory::GetForBrowserContext(profile)
+          ->GetDataStoreService();
+}
+
+FederatedLearningInternalsHandler::~FederatedLearningInternalsHandler() =
+    default;
+
+void FederatedLearningInternalsHandler::RegisterMessageCallbacks() {
+  RegisterMessageCallback(
+      "requestAdStoreInfo",
+      base::BindRepeating(
+          &FederatedLearningInternalsHandler::OnAdStoreRequestInfo,
+          base::Unretained(this)));
+}
+
+void FederatedLearningInternalsHandler::OnAdStoreRequestInfo(
+    const base::ListValue* args) {
+  auto* const ad_notification_data_store =
+      data_store_service_->GetAdNotificationTimingDataStore();
+
+  ad_notification_data_store
+      ->AsyncCall(&brave_federated::AdNotificationTimingDataStore::LoadLogs)
+      .Then(base::BindOnce(
+          &FederatedLearningInternalsHandler::OnAdStoreLogsReceivedCallback,
+          base::Unretained(this)));
```

Comment:
not safe to use Unretained like this, the handler may die way before the reply is received